File: shifai-ios/ShifAI/Presentation/Onboarding/OnboardingViewModel.swift
import Foundation

// MARK: - Onboarding ViewModel
// S3-1 to S3-5: Manages onboarding state, profile setup, and completion

final class OnboardingViewModel: ObservableObject {

    // S3-1: Welcome
    @Published var cycleDescription: String = ""

    // S3-2: Disclaimer
    @Published var disclaimerAccepted: Bool = false

    // S3-4: Profile Setup
    @Published var ageRange: String? = nil
    @Published var cycleLength: String? = nil
    @Published var selectedConditions: Set<String> = []
    @Published var trackedSymptoms: Set<String> = []

    // S3-5: Body Map
    @Published var markedZones: Set<BodyZone> = []
    @Published var hasMarkedBodyMap: Bool = false

    // MARK: - Condition-based Symptom Preselection

    var preselectedSymptoms: [String] {
        var symptoms = ["Crampes", "Fatigue", "Migraine", "Ballonnement", "Anxiété", "Insomnie"]

        if selectedConditions.contains("SOPK") {
            symptoms += ["Acné", "Chute de cheveux", "Cycle irrégulier", "Envies alimentaires"]
        }
        if selectedConditions.contains("Endométriose") {
            symptoms += ["Douleur pelvienne", "Mal de dos", "Nausée", "Diarrhée"]
        }

        return Array(Set(symptoms)).sorted()
    }

    // MARK: - Actions

    func toggleCondition(_ condition: String) {
        if selectedConditions.contains(condition) {
            selectedConditions.remove(condition)
        } else {
            // "Aucune" and "Je ne sais pas" are exclusive
            if condition == "Aucune" || condition == "Je ne sais pas" {
                selectedConditions = [condition]
            } else {
                selectedConditions.remove("Aucune")
                selectedConditions.remove("Je ne sais pas")
                selectedConditions.insert(condition)
            }
        }

        // Re-set tracked symptoms from preselection
        trackedSymptoms = Set(preselectedSymptoms)
    }

    func toggleTrackedSymptom(_ symptom: String) {
        if trackedSymptoms.contains(symptom) {
            trackedSymptoms.remove(symptom)
        } else {
            trackedSymptoms.insert(symptom)
        }
    }

    func toggleBodyZone(_ zone: BodyZone) {
        if markedZones.contains(zone) {
            markedZones.remove(zone)
        } else {
            markedZones.insert(zone)
        }
        hasMarkedBodyMap = !markedZones.isEmpty
    }

    // MARK: - Persistence

    func saveProfile() {
        let profile = UserProfile(
            id: "default",
            age: ageRangeToInt(ageRange),
            averageCycleLength: cycleLengthToInt(cycleLength),
            conditions: Array(selectedConditions),
            trackedSymptoms: Array(trackedSymptoms),
            locale: "fr",
            cycleDescription: cycleDescription,
            createdAt: Date(),
            updatedAt: Date()
        )

        // TODO: Save via UserProfileRepository when available
        UserDefaults.standard.set(true, forKey: "profileSetupComplete")
        print("Profile saved: \(profile)")
    }

    func completeOnboarding() {
        // Save body map data
        if !markedZones.isEmpty {
            let symptomRepo = SymptomRepository()
            let today = Date()
            for zone in markedZones {
                let log = SymptomLog(
                    id: UUID().uuidString,
                    date: today,
                    type: .pelvicPain,
                    intensity: 5,
                    bodyZone: zone,
                    painType: .cramping,
                    notes: "onboarding_first_log",
                    createdAt: today
                )
                try? symptomRepo.save(log)
            }
        }

        UserDefaults.standard.set(true, forKey: "onboardingComplete")
        UserDefaults.standard.set(Date(), forKey: "onboardingDate")

        // Schedule Quick Win J1 insight for immediate delivery
        QuickWinEngine.shared.scheduleJ1Insight()
    }

    // MARK: - Helpers

    private func ageRangeToInt(_ range: String?) -> Int? {
        switch range {
        case "18-24": return 21
        case "25-30": return 27
        case "31-35": return 33
        case "36-40": return 38
        case "41-45": return 43
        case "45+": return 48
        default: return nil
        }
    }

    private func cycleLengthToInt(_ length: String?) -> Int? {
        switch length {
        case "< 21 jours": return 19
        case "21-25": return 23
        case "26-30": return 28
        case "31-35": return 33
        case "35+": return 38
        default: return nil
        }
    }
}

File: shifai-android/app/src/main/java/com/shifai/presentation/onboarding/OnboardingViewModel.kt
package com.shifai.presentation.onboarding

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/**
 * Onboarding ViewModel — Android
 * S3-1 to S3-5: Manages 5-step onboarding state, profile setup, and completion
 *
 * Mirrors iOS OnboardingViewModel
 */
class OnboardingViewModel {

    // S3-1: Welcome
    private val _cycleDescription = MutableStateFlow("")
    val cycleDescription: StateFlow<String> = _cycleDescription.asStateFlow()

    // S3-2: Disclaimer
    private val _disclaimerAccepted = MutableStateFlow(false)
    val disclaimerAccepted: StateFlow<Boolean> = _disclaimerAccepted.asStateFlow()

    // S3-4: Profile Setup
    private val _ageRange = MutableStateFlow<String?>(null)
    val ageRange: StateFlow<String?> = _ageRange.asStateFlow()

    private val _cycleLength = MutableStateFlow<String?>(null)
    val cycleLength: StateFlow<String?> = _cycleLength.asStateFlow()

    private val _selectedConditions = MutableStateFlow<Set<String>>(emptySet())
    val selectedConditions: StateFlow<Set<String>> = _selectedConditions.asStateFlow()

    private val _trackedSymptoms = MutableStateFlow<Set<String>>(emptySet())
    val trackedSymptoms: StateFlow<Set<String>> = _trackedSymptoms.asStateFlow()

    // S3-5: Body Map
    data class BodyZoneEntry(val zone: String, val selected: Boolean = false)

    private val _markedZones = MutableStateFlow<Set<String>>(emptySet())
    val markedZones: StateFlow<Set<String>> = _markedZones.asStateFlow()

    // Preselected symptoms based on conditions
    val preselectedSymptoms: List<String>
        get() {
            val base = mutableListOf("Crampes", "Fatigue", "Migraine", "Ballonnement", "Anxiété", "Insomnie")
            val conditions = _selectedConditions.value

            if ("SOPK" in conditions) {
                base += listOf("Acné", "Chute de cheveux", "Cycle irrégulier", "Envies alimentaires")
            }
            if ("Endométriose" in conditions) {
                base += listOf("Douleur pelvienne", "Mal de dos", "Nausée", "Diarrhée")
            }

            return base.distinct().sorted()
        }

    // MARK: - Actions

    fun setCycleDescription(text: String) { _cycleDescription.value = text }
    fun toggleDisclaimer() { _disclaimerAccepted.value = !_disclaimerAccepted.value }
    fun setAgeRange(range: String) { _ageRange.value = range }
    fun setCycleLength(length: String) { _cycleLength.value = length }

    fun toggleCondition(condition: String) {
        val current = _selectedConditions.value.toMutableSet()
        if (condition in current) {
            current.remove(condition)
        } else {
            if (condition == "Aucune" || condition == "Je ne sais pas") {
                current.clear()
            } else {
                current.remove("Aucune")
                current.remove("Je ne sais pas")
            }
            current.add(condition)
        }
        _selectedConditions.value = current
        _trackedSymptoms.value = preselectedSymptoms.toSet()
    }

    fun toggleTrackedSymptom(symptom: String) {
        val current = _trackedSymptoms.value.toMutableSet()
        if (symptom in current) current.remove(symptom) else current.add(symptom)
        _trackedSymptoms.value = current
    }

    fun toggleBodyZone(zone: String) {
        val current = _markedZones.value.toMutableSet()
        if (zone in current) current.remove(zone) else current.add(zone)
        _markedZones.value = current
    }

    suspend fun saveProfile() {
        // TODO: Save via Room when DI is set up
    }

    suspend fun completeOnboarding() {
        // TODO: Save body map data and trigger J1 Quick Win
    }
}

/**
 * Quick Win Engine — Android
 * S3-6 (J1 Benchmark) + S3-7 (J3 Mini-Pattern) + S3-8 (Educational Drip)
 *
 * Mirrors iOS QuickWinEngine
 */
class QuickWinEngine {

    data class EducationalTip(
        val title: String,
        val body: String,
        val source: String?
    )

    val educationalTips = listOf(
        EducationalTip(
            "Phase menstruelle : le repos a du sens 🔴",
            "Pendant tes règles, le taux de progestérone et d'œstrogène chute. C'est normal de ressentir de la fatigue.",
            "ACOG"
        ),
        EducationalTip(
            "Stress et cycles : une connexion puissante 🧠",
            "Le cortisol peut retarder l'ovulation et allonger ton cycle. ShifAI va traquer cette corrélation.",
            "Harvard Health"
        ),
        EducationalTip(
            "Sommeil et hormones : un duo critique 😴",
            "La mélatonine influence directement la production de GnRH. Vise 7-8h régulières.",
            "Sleep Foundation"
        ),
        EducationalTip(
            "Phase folliculaire : ton énergie remonte 🌱",
            "Après les règles, les œstrogènes augmentent. C'est souvent le moment le plus dynamique.",
            "Clue"
        ),
        EducationalTip(
            "SOPK : comprendre les bases 💜",
            "Le SOPK touche 1 femme sur 10 — excès d'androgènes et résistance à l'insuline.",
            "WHO"
        ),
        EducationalTip(
            "Nutrition et cycle 🥗",
            "Les aliments anti-inflammatoires peuvent réduire les douleurs. En phase lutéale, +100-300 calories/jour.",
            "British Journal of Nutrition"
        ),
        EducationalTip(
            "Exercice et cycle 🏃‍♀️",
            "Folliculaire → HIIT/cardio. Lutéale → yoga/pilates. Écouter son énergie optimise les performances.",
            "British Journal of Sports Medicine"
        ),
        EducationalTip(
            "Endométriose : 7 ans pour un diagnostic ⏳",
            "Le suivi régulier de tes douleurs est l'un des meilleurs outils pour accélérer le diagnostic.",
            "Endometriosis UK"
        ),
        EducationalTip(
            "Phase ovulatoire : pic d'énergie 🌸",
            "Le pic de LH et d'œstrogène donne un boost d'énergie et de confiance.",
            "Healthline"
        ),
        EducationalTip(
            "Tu es unique 🌈",
            "Les normes sont des moyennes. ShifAI apprend TON rythme unique.",
            null
        ),
    )

    fun checkAndGenerateInsights(daysSinceOnboarding: Int) {
        when (daysSinceOnboarding) {
            0 -> generateJ1Benchmark()
            2 -> generateJ3MiniPattern()
            in 3..12 -> deliverEducationalTip(daysSinceOnboarding - 3)
        }
    }

    private fun generateJ1Benchmark() {
        // TODO: Create benchmark insight via Room DAO
    }

    private fun generateJ3MiniPattern() {
        // TODO: Analyze 3-day energy trend + symptom frequency via Room DAO
    }

    private fun deliverEducationalTip(index: Int) {
        if (index in educationalTips.indices) {
            val tip = educationalTips[index]
            // TODO: Save educational insight via Room DAO
        }
    }
}
